### PR TITLE
Dec 3 Merge

### PR DESCRIPTION
**Pull request by romechenko**
merged my Dec 2 work with your Dec 3

Starting my Dec 3

### CODE REVIEW

File: Qa.py [hunk 1]
```diff
@@ -29,47 +29,39 @@ def get_quarter_total(cur):
...
             continue
 
         # Get the percentage of students attempting n units for quarter
         s_count = get_count_student_attempts_n_units_for_quarter(cur, units)
-        # print "s_count {0:9} q_tot {1:9}".format(s_count, q_tots[itr])
         percent = (s_count/tot)*100
         persum += percent
 
-        print ("{0:<9.1f} {1:<9.4f}".format(
+        print ("{0:<9.0f} {1:<9.2f}".format(
```
Comment (romechenko) [inline Qa.py]: This modifies rounding of floats

File: Qa.py [hunk 2]
```diff
@@ -29,47 +29,39 @@ def get_quarter_total(cur):
     return float(cur.fetchone()[0])
 
 def get_count_student_attempts_n_units_for_quarter(cur, units):
-    cur.execute("SELECT COUNT(sid) "\
-                "FROM "\
+    cur.execute("SELECT COUNT(S.sid) "\
+                "FROM Student S, "\
```
Comment (romechenko) [inline Qa.py:33]: The cross product with S above results with a different count if it is not used here.

File: Qa.py [hunk 3]
```diff
@@ -29,47 +29,39 @@ def get_quarter_total(cur):
...
 
     #PRINT HEADER
-    print ("{0:<9} {1:<9}"
-        .format("Units", "Percentages"))
+    print ("{0:<9} {1:<15}".format("Units", "Percentages"))
 
     persum = 0;
 
-    for units in unitsPerQuarter:
+    for units in range(1, 21):
```
Comment (romechenko) [inline Qa.py:55]: Starting with 1 while less than 21

File: Qb.py [hunk 4]
```diff
@@ -10,8 +10,8 @@
 import time
 
 def connect():
-    user = os.environ['USER']
-    # user = "postgres"
+    # user = os.environ['USER']
+    user = "postgres"
```
Comment (romechenko) [inline Qb.py:14]: We need to make these consistent at the end.

File: Qb.py [hunk 5]
```diff
@@ -23,7 +23,7 @@ def get_average_gpa_student_attempts_n_units_for_quarter(cur, units):
                 "FROM "\
                     "(SELECT sid, SUM(s.units) as sum, AVG(grade) as GPA "\
                     "FROM Student s, Course c "\
-                    "WHERE s.cid = c.cid "\
+                    "WHERE s.cid = c.cid AND grade != 'nan'"\
```
Comment (romechenko) [inline Qb.py]: Non-letter grades may be NaN

File: Qc.py [hunk 6]
```diff
@@ -23,19 +23,20 @@ def get_average_gpa_best_worst_prof(cur):
...
                                 "SELECT instructor, AVG(grade) "\
                                 "FROM Meeting m, Student s "\
-                                "WHERE m.cid = s.cid "\
+                                "WHERE m.cid = s.cid AND s.grade != 'nan' "\
                                 "GROUP BY instructor"\
                             ") pag"\
                     ") mm "\
-                "WHERE avg = min OR avg = max;")
+                "WHERE avg = min OR avg = max "\
+                "GROUP BY avg, instructor;")
```
Comment (romechenko) [inline Qc.py:39]: In question d our grouping results in minimum values being listed before maximum values. Doing the group by operation here is for the sake of consistency.

File: Qc.py [hunk 7]
```diff
@@ -44,15 +45,14 @@ def main():
     cur = conn.cursor()
 
     #PRINT HEADER
-    print ("{0:<9}                       {1:<9}".format(
+    print ("{0:<32} {1:<13}".format(
```
Comment (romechenko) [inline Qc.py]: Professors have long names

File: Qc.py [hunk 8]
```diff
@@ -44,15 +45,14 @@ def main():
...
         "Instructor", "Average GPA"))
 
     # Get the percentage of students attempting n units for quarter
     results = get_average_gpa_best_worst_prof(cur)
-    # print "s_count {0:9} q_tot {1:9}".format(s_count, q_tots[itr])
 
-    for i in range(2):
-        print ("{0:<9} {1:<9.4f}".format(
+    for i in range(0, len(results)):
```
Comment (romechenko) [inline Qc.py:53]: There may be more than one professor with either best or worst GPA.

File: Qd.py [hunk 9]
```diff
@@ -25,22 +25,24 @@ def get_average_gpa_best_worst_prof_abc(cur):
...
                         "WHERE c.subject = 'ABC' AND crse>99 "\
                             "AND crse<200 AND c.cid=m.cid "\
-                            "AND m.cid=s.cid "\
+                            "AND m.cid=s.cid AND grade != 'nan' "\
+                            "AND instructor != 'null' "\
                         "GROUP BY crse, instructor"\
                      ") cpg, "\
                      "("\
-                        "SELECT crse, MIN(avg), MAX(avg) "\
+                        "SELECT crse, MIN(avg) as mingpa, MAX(avg) as maxgpa "\
```
Comment (romechenko) [inline Qd.py:33]: This may have been unnecesary

**Review by jpfischbein — APPROVED** [review]

**Pull request merged**